cupyx/scipy/ndimage/fourier.py

```diff
@@ -0,0 +1,194 @@
+        axis (int, optional): The axis of the real transform (only used when
+            ``n > -1``).
+        output (cupy.ndarray, optional):
+            If given, the result of shifting the input is placed in this array.
+
+    Returns:
+        output (cupy.ndarray): The filtered output.
+    """
+    ndim = input.ndim
+    output = _get_output_fourier(output, input)
+    axis = cupy.util._normalize_axis_index(axis, ndim)
+    sizes = filters._fix_sequence_arg(size, ndim, 'size')
+
+    output[...] = input
+    for ax, (size, ax_size) in enumerate(zip(sizes, output.shape)):
+
+        # compute the frequency grid in Hz
+        if ax == axis and n > 0:
+            arr = cupy.arange(ax_size, dtype=output.real.dtype)
+            arr /= n
+        else:
+            arr = cupy.fft.fftfreq(ax_size)
+        arr = arr.astype(output.real.dtype, copy=False)
+
+        # compute the uniform filter weights
+        arr *= size
+        cupy.sinc(arr, out=arr)
+
+        # reshape for broadcasting
+        arr = _reshape_nd(arr, ndim=ndim, axis=ax)
+        output *= arr
+
+    return output
+
+
+def fourier_shift(input, shift, n=-1, axis=-1, output=None):
+    """Multidimensional Fourier shift filter.
+
+    The array is multiplied with the Fourier transform of a shift operation.
+
+    Args:
+        input (cupy.ndarray): The input array. This should be in the Fourier
+            domain.
+        shift (float or sequence of float):  The size of shift. If a float,
+            `shift` is the same for all axes. If a sequence, `shift` has to
+            contain one value for each axis.
+        n (int, optional):  If `n` is negative (default), then the input is
+            assumed to be the result of a complex fft. If `n` is larger than or
+            equal to zero, the input is assumed to be the result of a real fft,
+            and `n` gives the length of the array before transformation along
+            the real transform direction.
+        axis (int, optional): The axis of the real transform (only used when
+            ``n > -1``).
+        output (cupy.ndarray, optional):
+            If given, the result of shifting the input is placed in this array.
+
+    Returns:
+        output (cupy.ndarray): The shifted output (in the Fourier domain).
+    """
+    ndim = input.ndim
+    output = _get_output_fourier_complex(output, input)
+    axis = cupy.util._normalize_axis_index(axis, ndim)
+    shifts = filters._fix_sequence_arg(shift, ndim, 'shift')
+
+    output[...] = input
+    for ax, (shiftk, ax_size) in enumerate(zip(shifts, output.shape)):
+        if shiftk == 0:
+            continue
+        if ax == axis and n > 0:
+            # cp.fft.rfftfreq(ax_size) * (-2j * np.pi * shiftk *  ax_size / n)
+            arr = cupy.arange(ax_size, dtype=output.dtype)
+            arr *= -2j * np.pi * shiftk / n
+        else:
+            arr = cupy.fft.fftfreq(ax_size)
+            arr = arr * (-2j * np.pi * shiftk)
```

Comment:
In-place won't work here because `arr` is real-valued prior to the multiplication, but complex afterwards.

We can make it slightly more efficient if we duplicate some internals of `fftfreq` rather than calling `fftfreq`, but I thought that was probably not worth the added complexity. I did do that for the real-valued case above where `rfftfreq` basically just corresponds to `arange`, so wasn't complicated to replace.